File: examples/simple/decimal-binary/python/_cott_impl/curriculum/decimal_binary/convert_binary_decimal.py
from cott_runtime import Err, Ok, Result

from curriculum.decimal_binary import binary_to_decimal, decimal_to_binary
from curriculum.decimal_binary_types import Conversion, ConversionError, ConversionResult, ConversionResult_Binary, ConversionResult_Decimal, Conversion_BinaryToDecimal, Conversion_DecimalToBinary


def convert_binary_decimal(operation: Conversion) -> Result[ConversionResult, ConversionError]:
    if isinstance(operation, Conversion_DecimalToBinary):
        binary = decimal_to_binary(operation.value)
        if isinstance(binary, Err):
            return Err(error=binary.error)
        return Ok(value=ConversionResult_Binary(digits=binary.value))
    assert isinstance(operation, Conversion_BinaryToDecimal)
    decimal = binary_to_decimal(operation.digits)
    if isinstance(decimal, Err):
        return Err(error=decimal.error)
    return Ok(value=ConversionResult_Decimal(value=decimal.value))

File: examples/kotlin/simple/decimal-binary/kotlin/cott_impl/curriculum/decimal_binary/convert_binary_decimal.kt
package cott_impl.curriculum.decimal_binary

internal fun convert_binary_decimal(operation: curriculum.decimal_binary.Conversion): cott_runtime.CottResult<curriculum.decimal_binary.ConversionResult, curriculum.decimal_binary.ConversionError> =
    when (operation) {
        is curriculum.decimal_binary.Conversion.DecimalToBinary ->
            when (val r = curriculum.decimal_binary.decimal_to_binary(operation.value)) {
                is cott_runtime.Ok -> cott_runtime.Ok(curriculum.decimal_binary.ConversionResult.Binary(r.value))
                is cott_runtime.Err -> cott_runtime.Err(r.error)
            }
        is curriculum.decimal_binary.Conversion.BinaryToDecimal ->
            when (val r = curriculum.decimal_binary.binary_to_decimal(operation.digits)) {
                is cott_runtime.Ok -> cott_runtime.Ok(curriculum.decimal_binary.ConversionResult.Decimal(r.value))
                is cott_runtime.Err -> cott_runtime.Err(r.error)
            }
    }
